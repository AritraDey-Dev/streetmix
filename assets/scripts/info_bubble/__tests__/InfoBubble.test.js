/* eslint-env jest */
import React from 'react'
import { fireEvent } from '@testing-library/react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import InfoBubble from '../InfoBubble'
import {
  BUILDING_LEFT_POSITION,
  BUILDING_RIGHT_POSITION
} from '../../segments/constants'

jest.mock('../../segments/view')
jest.mock('../../segments/buildings', () => {
  return {
    prettifyHeight: jest.fn(() => 10),
    // TODO: use mock building data
    BUILDINGS: {
      grass: {
        id: 'grass',
        label: 'Grass',
        spriteId: 'buildings--grass',
        hasFloors: false,
        sameOnBothSides: true
      }
    }
  }
})

const initialState = {
  infoBubble: {
    visible: false,
    descriptionVisible: false,
    mouseInside: false
  },
  ui: {
    activeSegment: 0
  },
  street: {
    segments: [],
    leftBuildingVariant: 'grass',
    rightBuildingVariant: 'grass'
  },
  system: {},
  locale: {
    locale: 'en'
  }
}

describe('InfoBubble', () => {
  it('renders', () => {
    const wrapper = renderWithReduxAndIntl(<InfoBubble />, { initialState })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  // TODO: this passes, but it doesn't render the <Description /> child component.
  it('shows description', () => {
    const wrapper = renderWithReduxAndIntl(<InfoBubble />, {
      initialState: {
        ...initialState,
        infoBubble: {
          visible: false,
          descriptionVisible: true,
          mouseInside: false
        }
      }
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('is visible', () => {
    const wrapper = renderWithReduxAndIntl(<InfoBubble />, {
      initialState: {
        ...initialState,
        infoBubble: {
          visible: true,
          descriptionVisible: false,
          mouseInside: false
        }
      }
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('shows building left info bubble', () => {
    const wrapper = renderWithReduxAndIntl(<InfoBubble />, {
      initialState: {
        ...initialState,
        ui: {
          activeSegment: BUILDING_LEFT_POSITION
        }
      }
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('shows building right info bubble', () => {
    const wrapper = renderWithReduxAndIntl(<InfoBubble />, {
      initialState: {
        ...initialState,
        ui: {
          activeSegment: BUILDING_RIGHT_POSITION
        }
      }
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  describe('interactions', () => {
    it('set info bubble mouse inside', () => {
      const { container, store } = renderWithReduxAndIntl(<InfoBubble />, {
        initialState
      })
      fireEvent.mouseEnter(container.firstChild)

      expect(store.getState().infoBubble.mouseInside).toEqual(true)
    })

    it('does not set info bubble mouse inside', () => {
      const { container, store } = renderWithReduxAndIntl(<InfoBubble />, {
        initialState
      })
      fireEvent.mouseLeave(container.firstChild)

      expect(store.getState().infoBubble.mouseInside).toEqual(false)
    })
  })
})
