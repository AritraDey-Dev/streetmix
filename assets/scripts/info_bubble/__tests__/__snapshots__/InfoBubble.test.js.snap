// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoBubble is visible 1`] = `
<div
  className="info-bubble info-bubble-type-segment visible"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <Triangle
    highlight={false}
  />
  <header>
    <Connect(EditableLabel)
      label={null}
      position={2}
      segment={Object {}}
    />
    <injectIntl(RemoveButton)
      segment={2}
    />
  </header>
  <div
    className="info-bubble-controls"
  >
    <IntlProvider
      defaultFormats={Object {}}
      defaultLocale="en"
      formats={Object {}}
      locale="en"
      messages={Object {}}
      onError={[Function]}
      textComponent={Symbol(react.fragment)}
    >
      <injectIntl(Connect(Variants))
        position={2}
        type={1}
      />
    </IntlProvider>
    <injectIntl(Connect(WidthControl))
      position={2}
    />
  </div>
  <Warnings
    segment={Object {}}
  />
  <Connect(Description)
    highlightTriangle={[Function]}
    infoBubbleEl={null}
    unhighlightTriangle={[Function]}
    updateBubbleDimensions={[Function]}
    updateHoverPolygon={[Function]}
  />
</div>
`;

exports[`InfoBubble renders 1`] = `
<div
  className="info-bubble info-bubble-type-segment"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <Triangle
    highlight={false}
  />
  <header>
    <Connect(EditableLabel)
      label={
        <IntlProvider
          defaultFormats={Object {}}
          defaultLocale="en"
          formats={Object {}}
          locale="en"
          messages={Object {}}
          onError={[Function]}
          textComponent={Symbol(react.fragment)}
        >
          <FormattedMessage
            defaultMessage="Unknown"
            id="segments.undefined"
            values={Object {}}
          />
        </IntlProvider>
      }
      position={0}
      segment={Object {}}
    />
    <injectIntl(RemoveButton)
      segment={0}
    />
  </header>
  <div
    className="info-bubble-controls"
  >
    <IntlProvider
      defaultFormats={Object {}}
      defaultLocale="en"
      formats={Object {}}
      locale="en"
      messages={Object {}}
      onError={[Function]}
      textComponent={Symbol(react.fragment)}
    >
      <injectIntl(Connect(Variants))
        position={0}
        type={1}
      />
    </IntlProvider>
    <injectIntl(Connect(WidthControl))
      position={0}
    />
  </div>
  <Warnings
    segment={Object {}}
  />
  <Connect(Description)
    highlightTriangle={[Function]}
    infoBubbleEl={null}
    unhighlightTriangle={[Function]}
    updateBubbleDimensions={[Function]}
    updateHoverPolygon={[Function]}
  />
</div>
`;

exports[`InfoBubble shows building left info bubble 1`] = `
<div
  className="info-bubble info-bubble-type-building"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <Triangle
    highlight={false}
  />
  <header>
    <Connect(EditableLabel)
      label={
        <IntlProvider
          defaultFormats={Object {}}
          defaultLocale="en"
          formats={Object {}}
          locale="en"
          messages={Object {}}
          onError={[Function]}
          textComponent={Symbol(react.fragment)}
        >
          <FormattedMessage
            defaultMessage="Grass"
            id="buildings.grass.name"
            values={Object {}}
          />
        </IntlProvider>
      }
      position="left"
      segment={Object {}}
    />
  </header>
  <div
    className="info-bubble-controls"
  >
    <IntlProvider
      defaultFormats={Object {}}
      defaultLocale="en"
      formats={Object {}}
      locale="en"
      messages={Object {}}
      onError={[Function]}
      textComponent={Symbol(react.fragment)}
    >
      <injectIntl(Connect(Variants))
        position="left"
        type={2}
      />
    </IntlProvider>
    <injectIntl(Connect(BuildingHeightControl))
      position="left"
    />
  </div>
  <Warnings
    segment={Object {}}
  />
  <Connect(Description)
    highlightTriangle={[Function]}
    infoBubbleEl={null}
    unhighlightTriangle={[Function]}
    updateBubbleDimensions={[Function]}
    updateHoverPolygon={[Function]}
  />
</div>
`;

exports[`InfoBubble shows building right info bubble 1`] = `
<div
  className="info-bubble info-bubble-type-building"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <Triangle
    highlight={false}
  />
  <header>
    <Connect(EditableLabel)
      label={
        <IntlProvider
          defaultFormats={Object {}}
          defaultLocale="en"
          formats={Object {}}
          locale="en"
          messages={Object {}}
          onError={[Function]}
          textComponent={Symbol(react.fragment)}
        >
          <FormattedMessage
            defaultMessage="Grass"
            id="buildings.grass.name"
            values={Object {}}
          />
        </IntlProvider>
      }
      position="right"
      segment={Object {}}
    />
  </header>
  <div
    className="info-bubble-controls"
  >
    <IntlProvider
      defaultFormats={Object {}}
      defaultLocale="en"
      formats={Object {}}
      locale="en"
      messages={Object {}}
      onError={[Function]}
      textComponent={Symbol(react.fragment)}
    >
      <injectIntl(Connect(Variants))
        position="right"
        type={3}
      />
    </IntlProvider>
    <injectIntl(Connect(BuildingHeightControl))
      position="right"
    />
  </div>
  <Warnings
    segment={Object {}}
  />
  <Connect(Description)
    highlightTriangle={[Function]}
    infoBubbleEl={null}
    unhighlightTriangle={[Function]}
    updateBubbleDimensions={[Function]}
    updateHoverPolygon={[Function]}
  />
</div>
`;

exports[`InfoBubble shows description 1`] = `
<div
  className="info-bubble info-bubble-type-segment show-description"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
>
  <Triangle
    highlight={false}
  />
  <header>
    <Connect(EditableLabel)
      label={null}
      position={1}
      segment={Object {}}
    />
    <injectIntl(RemoveButton)
      segment={1}
    />
  </header>
  <div
    className="info-bubble-controls"
  >
    <IntlProvider
      defaultFormats={Object {}}
      defaultLocale="en"
      formats={Object {}}
      locale="en"
      messages={Object {}}
      onError={[Function]}
      textComponent={Symbol(react.fragment)}
    >
      <injectIntl(Connect(Variants))
        position={1}
        type={1}
      />
    </IntlProvider>
    <injectIntl(Connect(WidthControl))
      position={1}
    />
  </div>
  <Warnings
    segment={Object {}}
  />
  <Connect(Description)
    highlightTriangle={[Function]}
    infoBubbleEl={null}
    unhighlightTriangle={[Function]}
    updateBubbleDimensions={[Function]}
    updateHoverPolygon={[Function]}
  />
</div>
`;
